### PR TITLE
Add limit offset support for cross shard queries

### PR DESCRIPTION
**Pull request by rafael**
### Description

This commit adds support to limit offset queries that have order by clauses eg:
```
   select id from users order by id limit 1 offset 1
```
Implementation modifies queries at runtime to send to all shards `LIMIT + OFFSET` and then relies on route primitive returning results in order. With this it calculates the proper limit offset that it needs to return in the response.

Giving this there is an important caveat that it's important to keep in mind:
Vitess enforces a max number of rows returned per query (defaults to 10k).
This effectively means that if a shard has more than `LIMIT + OFFSET` rows,
that query will fail.

However this should work for most of the common cases and makes Vitess more flexible
and SQL compatible.

### CODE REVIEW

File: data/test/vtgate/select_cases.txt [hunk 1]
```diff
@@ -460,6 +460,31 @@
...
+    "Opcode": "Limit",
+    "Count": 20,
+    "Offset": 10,
+    "Input": {
+      "Opcode": "SelectScatter",
+      "Keyspace": {
+        "Name": "user",
+        "Sharded": true
+      },
+      "Query": "select user_id from music order by user_id asc limit 20",
```
Comment (sougou) [inline data/test/vtgate/select_cases.txt]: This should be `limit 30`. However, given that 10 and 20 could be bind vars, the limit cannot be pre-computed at planbuilder time.

So, one option is to generate something `limit :__upper_limit`, and compute the value for this bind var as `count+limit` at execution time.

File: go/vt/vtgate/planbuilder/limit.go [hunk 2]
```diff
@@ -127,6 +123,15 @@ func (l *limit) SetLimit(limit *sqlparser.Limit) error {
 	}
 	l.elimit.Count = pv
 	l.input.SetUpperLimit(count)
```
Comment (sougou) [inline go/vt/vtgate/planbuilder/limit.go]: Here you should push the bind var instead.

File: go/vt/vtgate/engine/limit.go [hunk 3]
```diff
@@ -29,10 +29,10 @@ import (
 var _ Primitive = (*Limit)(nil)
 
 // Limit is a primitive that performs the LIMIT operation.
-// For now, it only supports count without offset.
 type Limit struct {
-	Count sqltypes.PlanValue
-	Input Primitive
+	Count  sqltypes.PlanValue
+	Offset *sqltypes.PlanValue
```
Comment (sougou) [inline go/vt/vtgate/engine/limit.go]: Better not to use a pointer. Use `Offset.IsNull` instead.

File: go/vt/vtgate/engine/limit.go [hunk 4]
```diff
@@ -56,12 +58,39 @@ func (l *Limit) Execute(vcursor VCursor, bindVars map[string]*querypb.BindVariab
 	if err != nil {
 		return nil, err
 	}
+	offset, err := l.fetchOffset(bindVars)
+	if err != nil {
+		return nil, err
+	}
+	// When offset is present, this bind var is set by the planner so we
+	// can hijack the limit value
+	if l.Offset != nil {
```
Comment (sougou) [inline go/vt/vtgate/engine/limit.go]: Use `IsNull`.

File: go/vt/vtgate/engine/limit_test.go [hunk 5]
```diff
@@ -93,6 +146,173 @@ func TestLimitExecute(t *testing.T) {
 	}
 }
 
+func TestLimitOffsetExecute(t *testing.T) {
```
Comment (sougou) [inline go/vt/vtgate/engine/limit_test.go]: I think it's better to add tests for those corner cases where count+offset==len, etc. That will give you the confidence on the off-by-one bugs, but will also prevent future bugs.

File: go/vt/vtgate/planbuilder/limit.go [hunk 6]
```diff
@@ -127,6 +123,18 @@ func (l *limit) SetLimit(limit *sqlparser.Limit) error {
...
+
+	offset, ok := limit.Offset.(*sqlparser.SQLVal)
+	if ok {
```
Comment (sougou) [inline go/vt/vtgate/planbuilder/limit.go]: Better to move this at the beginning and make the non-offset part as the else (or make this block return early).
Otherwise, we'll be calling `SetUpperLimit` twice.

File: go/vt/vtgate/planbuilder/postprocess.go [hunk 7]
```diff
@@ -140,6 +140,9 @@ func pushLimit(limit *sqlparser.Limit, bldr builder) (builder, error) {
 		return bldr, nil
 	}
 	lb := newLimit(bldr)
+	if limit != nil && limit.Offset != nil && len(orderBy) == 0 {
+		return nil, errors.New("unsupported: offset limit for cross-shard queries")
```
Comment (sougou) [inline go/vt/vtgate/planbuilder/postprocess.go]: This error should be `offset limit for cross-shard queries without order by` right?
I think if we want to enforce this, we should apply the rule even if offset is not specified.

I'm thinking we should not enforce this. If people don't want ordering, it's their problem.

**Comment by sougou** [comment]
I'm also thinking it's not worth special-casing the simple limit case for the bind var. I think the code will be simpler if we always pushed `__upper_limit` down.

File: go/vt/vtgate/planbuilder/limit.go [hunk 8]
```diff
@@ -126,7 +123,21 @@ func (l *limit) SetLimit(limit *sqlparser.Limit) error {
...
-	l.input.SetUpperLimit(count)
+
+	switch offset := limit.Offset.(type) {
+	case *sqlparser.SQLVal:
+		pv, err = sqlparser.NewPlanValue(offset)
+		if err != nil {
+			return err
+		}
+		l.elimit.Offset = pv
+	case nil:
```
Comment (sougou) [inline go/vt/vtgate/planbuilder/limit.go]: This case can be skipped.

File: go/vt/vtgate/planbuilder/postprocess.go [hunk 9]
```diff
@@ -130,7 +130,7 @@ func pushOrderBy(orderBy sqlparser.OrderBy, bldr builder) error {
 	return nil
 }
 
-func pushLimit(limit *sqlparser.Limit, bldr builder) (builder, error) {
+func pushLimit(limit *sqlparser.Limit, orderBy sqlparser.OrderBy, bldr builder) (builder, error) {
```
Comment (sougou) [inline go/vt/vtgate/planbuilder/postprocess.go]: No need to pass orderBy any more.

File: go/vt/vtgate/planbuilder/limit.go (hunk 8 above)
Comment (sougou) [reply go/vt/vtgate/planbuilder/limit.go]: Clarification: the case is needed, but it can be a no-op.

Comment (rafael) [reply go/vt/vtgate/planbuilder/limit.go]: done

**Comment by rafael** [comment]
Thanks @sougou !  Just addressed these last comments.
